Yosemite/Yosemite/Tools/POS/POSOrderService.swift

```diff
@@ -5,15 +5,79 @@ import class WooFoundation.CurrencyFormatter
 /// POSCartItem is different from the CartItem in the POS app layer.
 /// - The POS cart UI might show the cart items differently from how they appear in an order in wp-admin.
 public struct POSCartItem {
-    let product: POSItem
+    let item: any POSOrderableItem
     let quantity: Decimal
 
-    public init(product: POSItem, quantity: Decimal) {
-        self.product = product
+    public init(item: any POSOrderableItem, quantity: Decimal) {
+        self.item = item
         self.quantity = quantity
     }
 }
 
+extension [POSCartItem] {
+    public func matches(order: Order?) -> Bool {
+        guard let order else {
+            return self.isEmpty
+        }
+
+        let consolidatedCartItems = self.reduce(into: [POSCartItem]()) { partialResult, nextItem in
+            if let matchingIndex = partialResult.firstIndex(where: { $0.item.isEqual(to: nextItem.item) }) {
+                let itemToUpdate = partialResult[matchingIndex]
+                partialResult[matchingIndex] = POSCartItem(item: itemToUpdate.item, quantity: itemToUpdate.quantity + nextItem.quantity)
+            } else {
+                partialResult.append(nextItem)
+            }
+        }
+
+        let consolidatedOrderItems = order.items.reduce(into: [OrderItem]()) { partialResult, nextItem in
+            if let matchingIndex = partialResult.firstIndex(where: { $0.productID == nextItem.productID && $0.variationID == nextItem.variationID }) {
+                let itemToUpdate = partialResult[matchingIndex]
+                partialResult[matchingIndex] = itemToUpdate.copy(quantity: itemToUpdate.quantity + nextItem.quantity)
+            } else {
+                partialResult.append(nextItem)
+            }
+        }
+
+        guard consolidatedCartItems.count == consolidatedOrderItems.count else {
+            return false
+        }
+
+        for cartItem in consolidatedCartItems {
+            guard consolidatedOrderItems.contains(where: { orderItem in
+                cartItem.item.matches(orderItem: orderItem) && cartItem.quantity == orderItem.quantity
+            }) else {
+                return false
+            }
```

Comment:
nit: could this be
```suggestion
            consolidatedOrderItems.contains(where: { orderItem in
                cartItem.item.matches(orderItem: orderItem) && cartItem.quantity == orderItem.quantity
            })
```

Reply:
No, as that will never make this function return false when an orderItems array doesn't contain a matching cartItem.

We can't just return the result of whether it does contain that, because if we do we won't continue to check the other order items, we'd just return based on the first one we check.